go/ql/lib/semmle/go/security/UnicodeBypassValidationQuery.qll

```diff
@@ -0,0 +1,149 @@
+    (source instanceof UntrustedFlowSource or source instanceof Source) and
+    state instanceof PreValidation
+  }
+
+  override predicate isSanitizer(DataFlow::Node sanitizer, DataFlow::FlowState state) {
+    sanitizer instanceof Sanitizer and
+    state instanceof PostValidation
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node nodeFrom, DataFlow::FlowState stateFrom, DataFlow::Node nodeTo,
+    DataFlow::FlowState stateTo
+  ) {
+    (
+      exists(DataFlow::CallNode cn |
+        cn.getACalleeIncludingExternals().asFunction() instanceof EscapeFunction and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn.getResult()
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        (
+          cn.getACalleeIncludingExternals().asFunction() instanceof RegexpMatchFunction or
+          cn.getACalleeIncludingExternals().asFunction() instanceof RegexpReplaceFunction
+        ) and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn.getResult()
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        cn.getTarget().hasQualifiedName("strings", "Compare") and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = nodeFrom
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        cn.getTarget().hasQualifiedName("strings", "Cut") and
+        nodeFrom = cn.getArgument(0) and
+        nodeTo = cn.getResult([0, 1])
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        cn.getTarget().hasQualifiedName("strings", ["CutPrefix", "CutSuffix"]) and
+        nodeFrom = cn.getArgument(0) and
+        nodeTo = cn.getResult(0)
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        cn.getTarget()
+            .hasQualifiedName("strings",
+              [
+                "Fields", "FieldsFunc", "Replace", "ReplaceAll", "Split", "SplitAfter",
+                "SplitAfterN", "SplitN", "ToLower", "ToLowerSpecial", "ToTitle", "ToTitleSpecial",
+                "ToUpper", "ToUpperSpecial", "Trim", "TrimFunc", "TrimLeft", "TrimLeftFunc",
+                "TrimPrefix", "TrimRight", "TrimRightFunc", "TrimSpace", "TrimSuffix",
+              ]) and
+        nodeFrom = cn.getArgument(0) and
+        nodeTo = cn.getAResult()
+      )
+    ) and
+    stateFrom instanceof PreValidation and
+    stateTo instanceof PostValidation
+  }
+
+  /* A Unicode Tranformation (Unicode tranformation) is considered a sink when the algorithm used is either NFC or NFKC.  */
+  override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
+    (
+      exists(string unicodeNorm, DataFlow::MethodCallNode cn |
+        unicodeNorm = package("golang.org/x/text", "unicode/norm") and
+        cn.getTarget().hasQualifiedName(unicodeNorm, "Form", "String") and
+        sink = cn.getArgument(0)
+      )
+      or
+      sink instanceof Sink
+    ) and
+    (
+      state instanceof PostValidation
+      or
+      sink instanceof UntrustedUnicodeCharChecks and state instanceof PreValidation
```

Comment:
Since this is a bit tangly, add a comment here illustrating that we're looking for two different scenarios -- flow through something like `Replace("<", "_")`, and unicode normalisation guarded by a check like `if !strings.Contains("<", "_") { ... }`

Reply:
Next time, the comments would follow the query update progress, commit b3abf6c1d19e847f8b97f3b6cc8c70817eedf82c.